Update computing pillar (#186)
sa-tre/satre-specification: docs/source/pillars/computing_technology.md

# Computing technology

```{figure} ../../images/Capability_Map/full.drawio.svg
:alt: SATRE Pillars Capability Map
:align: center

SATRE Pillars Capability Map
```

This pillar concerns actions taken by the TRE operator to manage systems for storing, retrieving, analysing and sending information.

> Each TRE operator will have its own computing technology requirements.
> Requirements are drawn from information governance requirements which primarily specify the controls needed within the technical and computing infrastructure.
> Researcher personas influence the according to the technical knowledge and experience of researchers along with the work they need to perform within the system, a data scientist will have very different requirements to a clinician for example.

Reviewer: I think this paragraph might be aiming for something like this?

```suggestion
Each TRE operator will have its own computing technology requirements.
Information governance requirements primarily specify security controls needed within the technical and computing infrastructure.
Other computing technology requirements will be influenced by the technical knowledge and experience of those using the TRE, along with the work they need to perform within the system. 
For example, a data scientist will have very different requirements to a clinician.
```

Author: Are you happy with this @JimMadge ?

Reviewer 2: Sounds good to me :+1:.

Might just be tempted to remove some of the adjectives for brevity.

> Researcher personas influence the according to the technical knowledge and experience of researchers along with the work they need to perform within the system, a data scientist will have very different requirements to a clinician for example.

Author: I think the essential point is that requirements for this pillar come from: IG, what the TRE organisation can offer and the needs of researchers. Do you want to suggest a better phrasing?

The required compute resources will vary according to the scale of data and computational techniques employed during research.

<!-- See all pillars of the SATRE Pillars Capability Map here: {ref}`satre_pillars` -->

## End user computing

_The ability of the TRE operator to provide and manage devices, workspaces, interfaces and applications used by researchers to interact with underlying systems and data._

### End user computing interfaces

Software or system that allows people to interact with computers, applications, and services.

> This application group includes desktop, command line and code submission interface.

Reviewer 3: *interfaces* plural?

```{list-table}
:header-rows: 1
:name: tab-end-user-computing-interfaces

* - Statement
  - Guidance
  - Importance

* - You should disable the ability to copy data out of your TRE via the system clipboard.
  - A TRE user must not be able to copy sensitive data out of a workspace using the system clipboard.
    A TRE may allow user to paste text into a workspace.

>   - Recommended

Author: Feels like a "Mandatory" to me. What do you think?

>   - Recommended

Reviewer 2: I would say so. It feels like it is a big risk for not much benefit. Maybe TREs who are much more confident about safe people are less concerned here. It might be that we came to recommended here to not be too prescriptive.

In any case, I don't think we can say "should not" in the statement and then "must not" in the guidance.

@manics do you remember any extra background context here?

  - TREs which allow users to connect from their own devices should not require the installation of any bespoke TRE application on the user's device.
    In practice a web browser is the most common way to achieve this.

* - You should disable the ability to copy data out of your TRE via the system clipboard.
  - A TRE user must not be able to copy sensitive data out of a workspace using the system clipboard.
    A TRE may allow user to paste text into a workspace.
  - Recommended
* - Your TRE workspace should provide an environment familiar to your users.
  - This may take the form of a virtual Windows or Linux desktops, non-desktop interfaces such as JupyterLab and other web applications, or a terminal.
    Bespoke TRE-specific software should be avoided when widely used alternatives already exist.
  - Recommended
* - Your TRE should take accessibility for users with disabilities into account.
  - The restricted nature of TREs means many assistive tools such as screenreaders in a virtual desktop may not be allowed, but other options such as colour schemes, font sizes, and resizing user interface elements, should be supported.

>   - Recommended

Author: I think @machintim's suggestion is that this should be a SATRE-level principle that applies across everything, not a sentence under a single capability.

>   - Recommended

Author: Moved into principles. @harisood to rework these based on recent Collaboration Cafe.

Each TRE operator will have its own computing technology requirements.
The security controls needed by the computing infrastructure will depend on information governance requirements.
Other computing requirements will be influenced by the technical knowledge and experience of those using the TRE, along with the work they need to perform within the system. 
For example, a data scientist will have very different requirements to a clinician.
The required compute resources will vary according to the scale of data and computational techniques employed during research.

<!-- See all pillars of the SATRE Pillars Capability Map here: {ref}`satre_pillars` -->

## End user computing

The ability of the TRE operator to provide and manage devices, workspaces, interfaces and applications used by researchers to interact with underlying systems and data.

### End user computing interfaces

Software or systems that allows people to interact with the TRE.
This may include desktop, command-line and/or code-submission interfaces.

```{list-table}
:header-rows: 1
:name: tab-end-user-computing-interfaces

* - Statement
  - Guidance
  - Importance

> * - You should disable the ability to copy data out of your TRE via the system clipboard.
>   - A TRE user must not be able to copy sensitive data out of a workspace using the system clipboard.
>     A TRE may allow user to paste text into a workspace.
>   - Recommended

Author: ```suggestion
* - You must not allow users to copy data out of your TRE via the system clipboard.
  - A TRE user must not be able to copy sensitive data out of a workspace using the system clipboard.
    A TRE may allow user to paste text into a workspace.
    This might not be relevant to your TRE, for example if your user interface does not have a clipboard.
  - Required
```

* - Statement
  - Guidance
  - Importance
* - Your TRE must provide software applications that are relevant to working with the data in the TRE.
  - The tools provided will depend on the types of data in the TRE, and the expectations of users of the TRE.
    For users working in a TRE via a virtual desktop, this may include programming languages such as Python and R, integrated development environments, Jupyter notebooks, office type applications such as word processors and spreadsheets, command line tools, etc.
    TREs with non-desktop interfaces should similarly consider carefully which applications are best suited for the researchers needs when interacting with the data, for example "point and click" GUI tools for querying a database and generating plots of data.
    The set of tools should be reviewed regularly to ensure they are up to date.

>   - Mandatory

Reviewer 3: Does the specification need an appendix where we could put examples of TRE components? This could read something like:

- Your TRE must provide software applications that are relevant to working with the data in the TRE.
- The set of tools provided will depend on the types of data in the TRE, and the expectations of the TRE users, and should be reviewed regularly to ensure they are up to date. For examples of the kinds of software applications used by existing TREs, refer to the [appendix] <- *link to relevant part*
- Mandatory

>   - Mandatory

Author: Where things have been removed, I'm not completely sure why. I think it's fine to say that we choose to keep them (and amend the WP3 document accordingly). @JimMadge: do you want to reword this to be a bit less verbose?

>   - Mandatory

Author: Reverted in be1f53a

    TREs where the analysis code is developed on the access machine (as opppose to within the TRE) should provide documentation detailing the mechanism by which code is submitted to the TRE.
  - Mandatory
* - Your TRE should, where possible, automatically apply security related updates for user software.
  - Reducing the risk of exploitable vulnerabilities in installed software will increase the security of your TRE.
  - Recommended
* - Your TRE could provide shared services that are accessible to users in the same project.
  - This may include shared file storage, databases, collaborative writing, and other web applications.
    This must only be shared amongst users within the same project.
  - Optional
* - Your TRE could include licenced commercial software if required by researchers, but additional risks must be recorded and mitigated where neccesary.
  - For example, if an application must connect to an external licensing server, you must be confident that only licensing information is sent to this server, and that any network connections are secure.
  - Optional
```

> ### Code Version Control System

Reviewer 3: ```suggestion
### Enabling Reproducible Data Analysis and Version Control of Research Code
```

Author: I don't love the current name but I think the point is that it's an "application" in architecture-language so the name needs to be a description of an object/system.

Reviewer 3: How about that? Feel free to reject entirely

Author: It's not wrong, but it's a bit verbose! :joy:

Reviewer 3: how about now?

Author: I'm going to ignore this to stay in-sync with WP3, but please comment on the doc in teams or discuss with @machintim when he gets back.

An application that provides version control and collaboration features for code developed inside the TRE.

```{list-table}
:header-rows: 1
:name: tab-end-user-code-vcs

* - Statement
  - Guidance
  - Importance
* - Your TRE should provide tools to encourage best-practice in reproducibly analysing data.
  - Reproducibility of analyses improves auditability and accountability of how data has been used, as well as being best-practice in research.
    This may include version control software, and tools for developing and running data analysis pipelines.
  - Recommended

> ```

Reviewer 3: They aren't the same thing, but I do think here the sentiment is that version control helps with reproducible data analyses.

I think this section is good but I agree with Jim that the header should probably be renamed

> ```

Author: If you take a look at the diagram in WP3 the point here is that a VCS system is an Application Component of the End-User-Computing capability. We need to say something sensible about **why** a VCS system is a useful component, but it doesn't have to be this. Thoughts @JimMadge/@edwardchalstrey1?

> ```

Author: I'm going to ignore this to stay in-sync with WP3, but please comment on the doc in teams or discuss with @machintim when he gets back.

### Artefact management

> A service that manages and organizes software artefacts such as code libraries, dependencies, and build outputs in a centralised repository.

Author: Again, a non-ideal name for the Application Component that provides access to external software. What would you call it @JimMadge ?

```{list-table}
:header-rows: 1
:name: tab-end-user-artefact-management

* - Statement
  - Guidance
  - Importance
* - Your TRE could provide limited access to some public software repositories or container registries.
  - For example, a TRE may allow installation of packages from Python or R repositories, or provide an internal mirror with approved packages.
    Similarly a subset of public containers could be made available, or individual container images via an internal container registry.
  - Optional
* - You could choose to tightly control access to public repositories.
  - This might involve only allowing a limited selection of allowed repositories.
    You might also choose to scan for malicious packages and/or go through an approval process before allowing code into the technical environment.

>   - Optional

Author: Agree, these could be merged. Can you make a suggestion to do so?

  - If a TRE supports users conducting computationally intensive research it should provide access to dynamically scaleable compute or the equivalent.
    For example this may be in the form of a batch scheduler on a HPC cluster, or a dynamically created compute nodes on a cloud platform.
  - Recommended
* - Your TRE should be able to provide access to accelerators such as GPUs if required by users.
  - GPUs and other accelerators are commonly used in machine learning and other computationally intensive research.
    TREs should make it clear to users whether GPUs and other resources are available whilst projects are being assessed.

>   - Recommended

Author: This is one of the things to discuss. I'm also in favour of keeping it.

* - Your TRE could integrate with large-scale data analytics tools for working with large datasets.
  - For example, Spark and Hadoop can be used for distributed computing across a cluster.
    This may be an advantage where a TRE is using an amount of data that is too large for single-machine computing to be practical.
  - Optional
```

## Infrastructure management

The ability of the TRE operator to instantiate, deploy, change or remove physical or virtual infrastructure.

### Infrastructure deployment

The process of setting up and configuring infrastructure components and resources to support applications or services.

> This involves development, installation, configuration, and validation.

Author: Are those different Processes? Current processes here are:

- Infrastructure Deployment
- Change process
- Infrastructure Decommissing
- Availability Management

How would you prefer to split this?

    Consider the costs and practicality of whether this will work for your situation.
  - Recommended
```

### Infrastructure decommissioning

The process of retiring or removing infrastructure assets that are no longer needed or outdated, ensuring proper data handling and disposal.

```{list-table}
:header-rows: 1
:name: tab-infrastructure-decommissioning

* - Statement
  - Guidance
  - Importance
* - You must have a documented procedure for removing infrastructure when it is no longer needed
  - Removing unused infrastructure not only reduces costs and management burden but also reduces the attack surface of a TRE and reduces the risk of unaddressed vulnerabilities.
  - Mandatory
```

### Availability management

The process of ensuring all IT infrastructure meets the agreed levels of availability.

```{list-table}
:header-rows: 1
:name: tab-infrastructure-availability

* - Statement
  - Guidance
  - Importance
* - You should understand the availability and uptime guarantees of any providers that you rely on.
  - For remote TREs this might include your cloud provider(s) and/or data centre operators.

>     For on-premises TREs, it might be worth considering your ISP and electricity provider.

Reviewer 2: Is the electricity provider suggestion based on experience?

Definitely not my area but I'm not aware of uptime guarantees from utilities. I would think your primary concern here is that you have enough UPS capacity to safely shut everything down if there is a power cut.

Author: Yeah, this seems weird to me. Feels like the kind of thing you should mitigate (eg. with UPS as you suggest) rather than obtain a third-party guarantee. Can you reword to make this point?

Author: ```suggestion
    For on-premises TREs, it might be worth using an uninterruptable power supply (UPS) and planning how you would deal with internet outages.
```

  - Recommended
* - You should develop an availability target or statement and share this with your users.
  - Understanding how and when the TRE might be unavailable will help your projects in planning their work.
  - Recommended
```

### Network management

> An application used to manage and control network devices, ensuring proper functioning, security, and performance of the network.

Reviewer 2: ```suggestion
Managing network infrastructure, ensuring proper functioning, security, and performance.
```

Author: I think @machintim wants to make it clear which things are processes/applications/data assets. Can you reword in a way that isn't ambiguous?

Author: ```suggestion
An application used to manage network infrastructure, ensuring proper functioning, security, and performance.
```

* - Statement
  - Guidance
  - Importance
* - Your TRE must maintain segregation of users and data from different projects when using non-standard compute.
  - High performance or specialist compute is often shared amongst multiple users.
    Users and data must remain segregated at all times.

>     For example, when using physical compute resources, all sensitive data could be securely wiped before another user is given access to that same node.

Reviewer 4: Is wording on could be strong enough here, if we are saying it's mandatory

Author: - Keep users and data separated (this is mandatory)
- All sensitive data could be securely wiped (one possible way of achieving the above)

  - Network infrastructure must prevent unauthorised access to resources on the network.
    This may include firewalls, network segmentation, and restricting connections to the network.
  - Mandatory
* - You must monitor the network configuration of your TRE to check for misconfigurations and vulnerabilities.
  - This may include regular vulnerability scanning, and penetration testing.

>   - Mandatory

Reviewer 2: This feels like an odd one to remove. How can we be confident a TRE is correctly configured, and secure, if we don't test that?

I think this covers more than just networking though.

Author: I agree.

Author: Reverted in 39cbbe0

  - Mandatory
```

## Capacity management

### Capacity planning

The process of forecasting and determining the resources required to meet the demands of an application or system, ensuring that adequate resources are available when needed.

```{list-table}
:header-rows: 1
:name: tab-infrastructure-capacity

* - Statement
  - Guidance
  - Importance
* - You must ensure that all projects understand what resources are available and what the associated costs will be before the project starts.
  - For on-premises systems this might be related to the available hardware, for cloud-based systems there might be limits on how many instances of a particular resource (_e.g._ GPUs) can be used. Projects should use this information to understand whether the available resources will be sufficient for their requirements.
  - Mandatory
* - You should ensure that the anticipated needs of projects can be satisfied using available resources.
  - Note that this does not require you to accept requests for additional resources, but rather that promises made about resource availability before a project starts should be honoured wherever possible.
  - Recommended

> * - You must have a procedure for increasing/decreasing available resources.

Reviewer 3: ```suggestion
* - TRES with the ability to increase/decrease available resources must have a documented procedure in place for how to do this.
```

Author: A bit of a different tack, but after re-reading the guidance, I think the following might be what we're trying to say here:

```suggestion
* - You must have a procedure for allocating available resources among projects.
```

  - For cloud-based TREs this may involve scaling resources, such as virtual machines or databases, or deploying additional resources.
    For on-premises TREs this may involve a procurement process to ensure that necessary resources are available.
    Not all requests for capacity increase must necessarily be granted, but having a clear process will help projects understand when/why/how they can make use of additional capacity.
  - Mandatory
```

### Billing

The process of generating and managing invoices and bills for projects within the TRE.
It involves calculation, issuance, and recording of payments and receipts.

```{list-table}
:header-rows: 1
:name: tab-infrastructure-billing

* - Statement
  - Guidance
  - Importance
* - You must ensure that the anticipated resource requirements will not result in overspending by the TRE.
  - For cloud-based TREs this may involve budgeting and/or restricting resource consumption on a project-by-project basis.
    For on-premises TREs this may involve managing expectations to match the available resource.
  - Mandatory
```

> ### Configuration management

Reviewer 2: This section feels in the wrong place. Shouldn't it be under (or alongside) infrastructure management?

Author: The idea here is that billing is part of managing capacity. The only statement we had about billing was to do with infrastructure, but we could/should add something about personpower here too. Would that make it more coherent for you @JimMadge ?

Author: One to raise with @machintim when he gets back.
